File: proposed-projects/round-trip-experiment/reverse/20260909T175143Z_62b61540/20260910T094302Z_8f8be025/ported_implementation/gbnf/utils/errors/input_parse_error.py
from .build_error_position import build_error_position
from .errors_types import ValidInput
from .get_input_as_string import get_input_as_string

INPUT_PARSER_ERROR_HEADER_MESSAGE = "Failed to parse input string:"


def _build_message(src: str, pos: int) -> str:
    return "\n".join(
        [
            INPUT_PARSER_ERROR_HEADER_MESSAGE,
            "",
            *build_error_position(src, pos),
        ]
    )


class InputParseError(Exception):
    def __init__(
        self,
        most_recent_input: ValidInput,
        pos: int,
        previous_input: ValidInput = "",
    ):
        super().__init__(
            _build_message(
                f"{get_input_as_string(previous_input)}"
                f"{get_input_as_string(most_recent_input)}",
                pos + len(get_input_as_string(previous_input)),
            )
        )
        self.name = "InputParseError"
        self.most_recent_input = most_recent_input
        self.pos = pos
        self.previous_input = previous_input

    @property
    def src(self) -> str:
        return (
            f"{get_input_as_string(self.previous_input)}"
            f"{get_input_as_string(self.most_recent_input)}"
        )

    @property
    def error_for_most_recent_input(self) -> str:
        return _build_message(get_input_as_string(self.most_recent_input), self.pos)

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, InputParseError):
            return NotImplemented
        return self.pos == other.pos and str(self) == str(other)

    def __hash__(self) -> int:
        return hash((self.pos, str(self)))

File: packages/gbnf-experiment/data/20260909T180307Z_280f3890/ported_implementation/src/utils/errors/input-parse-error.ts
import { buildErrorPosition } from './build-error-position.js';
import type { ValidInput } from './errors-types.js';
import { getInputAsString } from './get-input-as-string.js';

export const INPUT_PARSER_ERROR_HEADER_MESSAGE = 'Failed to parse input string:';

const buildMessage = (src: string, pos: number): string =>
  [INPUT_PARSER_ERROR_HEADER_MESSAGE, '', ...buildErrorPosition(src, pos)].join('\n');

export class InputParseError extends Error {
  readonly mostRecentInput: ValidInput;
  readonly pos: number;
  readonly previousInput: ValidInput;

  constructor(mostRecentInput: ValidInput, pos: number, previousInput: ValidInput = '') {
    const previous = getInputAsString(previousInput);
    super(buildMessage(`${previous}${getInputAsString(mostRecentInput)}`, pos + previous.length));
    Object.setPrototypeOf(this, InputParseError.prototype);
    this.name = 'InputParseError';
    this.mostRecentInput = mostRecentInput;
    this.pos = pos;
    this.previousInput = previousInput;
  }

  get src(): string {
    return `${getInputAsString(this.previousInput)}${getInputAsString(this.mostRecentInput)}`;
  }

  get errorForMostRecentInput(): string {
    return buildMessage(getInputAsString(this.mostRecentInput), this.pos);
  }
}
